src/twisted/internet/test/test_inlinecb.py

```diff
@@ -197,6 +224,51 @@ def calling():
         self.assertIn("Error Marker", tb)
         self.assertIn("in erroring", f.getTraceback())
 
+    @skipIf(sys.version_info < (3, 13), "new in Python 3.13")
+    def test_forwardLotsOfTracebacks_313(self):
+        """
+        Several Chained inlineCallbacks gives information about all generators.
+
+        A wider test with 4 chained inline callbacks.
+
+        Application stack-trace should be reported for Python 3.13, but not
+        earlier versions.
+
+        Note that the previous test is testing the simple case, and this one is
+        testing the deep recursion case.
+
+        That case needs specific code in failure.py to accomodate to stack
+        breakage introduced by throwExceptionIntoGenerator.
+
+        Hence we keep the two tests in order to sort out which code we
+        might have regression in.
+        """
+
+        @inlineCallbacks
+        def erroring():
+            yield "forcing generator"
+            raise Exception("Error Marker")
+
+        @inlineCallbacks
+        def calling3():
+            yield erroring()
+
+        @inlineCallbacks
+        def calling2():
+            yield calling3()
+
+        @inlineCallbacks
+        def calling():
+            yield calling2()
+
+        d = calling()
+        f = self.failureResultOf(d)
+        tb = f.getTraceback()
+        self.assertIn("in calling", tb)
+        self.assertIn("yield calling2", tb)
```

Comment:
what happens in Python 3.13, why we don't see the trace for `calling3`?

Reply:
I think this is just another change in traceback formatting (like the last 2 python releases), but not indicative of a change in behavior.  When running it with a debugger, I see the entire callback chain, callback3 is just missing from the exception text. 

`<pre>  ForwardTraceBackTests
    test_forwardLotsOfTracebacks_312 ...                              <span style="color:#5555FF"><b>[SKIPPED]</b></span>
    test_forwardLotsOfTracebacks_313 ... Jumping into debugger for post-mortem of exception &apos;Error Marker&apos;:
&gt; /home/eevel/twisted/src/twisted/internet/test/test_inlinecb.py(251)erroring()
-&gt; raise Exception(&quot;Error Marker&quot;)
(Pdb) n
Jumping into debugger for post-mortem of exception &apos;Error Marker&apos;:
&gt; /home/eevel/twisted/src/twisted/internet/test/test_inlinecb.py(255)calling3()
-&gt; yield erroring()
(Pdb) n
Jumping into debugger for post-mortem of exception &apos;Error Marker&apos;:
&gt; /home/eevel/twisted/src/twisted/internet/test/test_inlinecb.py(259)calling2()
-&gt; yield calling3()
(Pdb) n
Jumping into debugger for post-mortem of exception &apos;Error Marker&apos;:
&gt; /home/eevel/twisted/src/twisted/internet/test/test_inlinecb.py(263)calling()
-&gt; yield calling2()
(Pdb) n
&gt; /home/eevel/twisted/src/twisted/internet/test/test_inlinecb.py(269)test_forwardLotsOfTracebacks_313()
-&gt; tb = f.getTraceback()
(Pdb) n
&gt; /home/eevel/twisted/src/twisted/internet/test/test_inlinecb.py(270)test_forwardLotsOfTracebacks_313()
-&gt; self.assertIn(&quot;in calling&quot;, tb)
(Pdb) tb
&apos;Traceback (most recent call last):\n  File &quot;/home/eevel/twisted/src/twisted/python/util.py&quot;, line 955, in runWithWarningsSuppressed\n    return f(*args, **kwargs)\n  File &quot;/home/eevel/twisted/src/twisted/internet/test/test_inlinecb.py&quot;, line 265, in test_forwardLotsOfTracebacks_313\n    d = calling()\n  File &quot;/home/eevel/twisted/src/twisted/internet/defer.py&quot;, line 2260, in unwindGenerator\n    return _cancellableInlineCallbacks(gen)\n  File &quot;/home/eevel/twisted/src/twisted/internet/defer.py&quot;, line 2172, in _cancellableInlineCallbacks\n    _inlineCallbacks(None, gen, status, _copy_context())\n--- &lt;exception caught here&gt; ---\n  File &quot;/home/eevel/twisted/src/twisted/internet/defer.py&quot;, line 1999, in _inlineCallbacks\n    result = context.run(\n  File &quot;/home/eevel/twisted/src/twisted/python/failure.py&quot;, line 519, in throwExceptionIntoGenerator\n    return g.throw(self.value.with_traceback(self.tb))\n  File &quot;/home/eevel/twisted/src/twisted/internet/test/test_inlinecb.py&quot;, line 263, in calling\n    yield calling2()\nbuiltins.Exception: Error Marker\n&apos;
</pre>`

This is the traceback, cleaned up for visibility:
`'Traceback (most recent call last):\n
  File "/home/eevel/twisted/src/twisted/python/util.py", line 955, in runWithWarningsSuppressed\n
    return f(*args, **kwargs)\n
  File "/home/eevel/twisted/src/twisted/internet/test/test_inlinecb.py", line 265, in test_forwardLotsOfTracebacks_313\n
    d = calling()\n  File "/home/eevel/twisted/src/twisted/internet/defer.py", line 2260, in unwindGenerator\n
    return _cancellableInlineCallbacks(gen)\n
  File "/home/eevel/twisted/src/twisted/internet/defer.py", line 2172, in _cancellableInlineCallbacks\n
    _inlineCallbacks(None, gen, status, _copy_context())\n
--- <exception caught here> ---\n
  File "/home/eevel/twisted/src/twisted/internet/defer.py", line 1999, in _inlineCallbacks\n
    result = context.run(\n
  File "/home/eevel/twisted/src/twisted/python/failure.py", line 519, in throwExceptionIntoGenerator\n
    return g.throw(self.value.with_traceback(self.tb))\n
  File "/home/eevel/twisted/src/twisted/internet/test/test_inlinecb.py", line 263, in calling\n
    yield calling2()\nbuiltins.Exception: Error Marker\n'`

Reply:
I'm open to suggestions for how to change this test, but I don't see a way to make this check more effective solely by asserting on the traceback contents.

Reply:
I've added an assertion that callback3 is missing, and updated the docstring to reflect this.

Reply:
I have reformated your example.
I am not sure what is going on here.
I see that the `erroring` call is also missing.

It looks like Traceback now only shows the top 2 stacks... I don't know.

I don't see anyting about this mentioned in the release notes https://docs.python.org/3.13/whatsnew/3.13.html

Maybe this is a bug in Python 3.13 ?

Reply:
Still investigating this one.

Reply:
Still digging, but I think this might be related to Traceback.exc_type vs Traceback.exc_type_str.

Reply:
I am ruling out TracebackException.exc_type vs exc_type_str as a cause.  After digging around in pdb, the codepath appears to be identical, as are the results, until it starts retrieving lines from py3/lib/linecache.getline().  Here are some screenshots of the BdbQuit traceback in 3.12 and 3.13.  The 3.13 tb is more sparse overall, and color-highlighted. 

Python 3.12:
![Screenshot_2024-02-28_22-16-50](https://github.com/twisted/twisted/assets/5155920/67325810-877a-4a44-af53-9d0034360a0e)

Python 3.13:
![Screenshot_2024-02-28_22-16-16](https://github.com/twisted/twisted/assets/5155920/fd74d066-7e52-4ca3-aa50-20d46ab4e561)

Reply:
I've updated the docstring for the 3.13 test to make the reason for this change more apparent.